### PR TITLE
Generalize userid authentication

### PR DESCRIPTION
**Pull request by harbu**
## Summary

Please provide a summary of the changes and a motivation if applicable.

## Changes

Provide a bullet list of individual changes. Leave this section out if change
set is small and obvious from summary.

## TODO

- [ ] Change 4*`randomUserId` functions to use random length instead of Ethereum-compatible 20 bytes
- [x] Make a separate PR about removing legacy key support, can merge "immediately"
- [x] Move (branded) `type UserID` type from `dht` to `trackerless-network`
- [ ] Grep `address` and `userAddress` and replace with e.g. `userId` in project.
- [x] Test function `randomUserId()` for tests
- [ ] Is the `userAddress` field deprecated in The Graph? Should we use the `userId` field, or both?
- [ ] Could we have a smart `toEthereumAddress` utility so that we wouldn't need `isEthereumAddressByteArray` function?

## Checklist before requesting a review

- [ ] Is this a breaking change? If it is, be clear in summary.
- [ ] Read through code myself one more time.
- [ ] Make sure any and all `TODO` comments left behind are meant to be left in.
- [ ] Has reasonable passing test coverage?
- [ ] Updated changelog if applicable.
- [ ] Updated documentation if applicable.

### CODE REVIEW

File: packages/sdk/src/contracts/StreamRegistry.ts [hunk 1]
```diff
@@ -356,9 +358,9 @@ export class StreamRegistry {
         if (isPublicPermissionQuery(query)) {
             return this.streamRegistryContractReadonly.hasPublicPermission(streamId, permissionType)
         } else if (query.allowPublic) {
-            return this.streamRegistryContractReadonly.hasPermission(streamId, toEthereumAddress(query.user), permissionType)
+            return this.streamRegistryContractReadonly.hasPermission(streamId, binaryToHex(query.user), permissionType)
```
Comment (harbu) [inline packages/sdk/src/contracts/StreamRegistry.ts]: is `binaryToHex` needed?

File: packages/sdk/src/contracts/StreamRegistry.ts [hunk 2]
```diff
@@ -356,9 +358,9 @@ export class StreamRegistry {
...
         } else {
-            return this.streamRegistryContractReadonly.hasDirectPermission(streamId, toEthereumAddress(query.user), permissionType)
+            return this.streamRegistryContractReadonly.hasDirectPermission(streamId, binaryToHex(query.user), permissionType)
```
Comment (harbu) [inline packages/sdk/src/contracts/StreamRegistry.ts]: is `binaryToHex` needed?

File: packages/sdk/src/contracts/StreamRegistry.ts [hunk 3]
```diff
@@ -485,17 +487,19 @@ export class StreamRegistry {
         await waitForTx(txToSubmit)
     }
 
-    private async isStreamPublisher_nonCached(streamId: StreamID, userAddress: EthereumAddress): Promise<boolean> {
+    private async isStreamPublisher_nonCached(streamId: StreamID, userId: Uint8Array): Promise<boolean> {
         try {
-            return await this.streamRegistryContractReadonly.hasPermission(streamId, userAddress, streamPermissionToSolidityType(StreamPermission.PUBLISH))
+            const permission = await this.streamRegistryContractReadonly.getPermissionsForUserId(streamId, userId)
```
Comment (harbu) [inline packages/sdk/src/contracts/StreamRegistry.ts]: Consider renaming to 

`permissionsRecord`

File: packages/sdk/src/contracts/StreamRegistry.ts [hunk 4]
```diff
@@ -485,17 +487,19 @@ export class StreamRegistry {
...
         } catch (err) {
             return streamContractErrorProcessor(err, streamId, 'StreamPermission')
         }
     }
 
-    private async isStreamSubscriber_nonCached(streamId: StreamID, userAddress: EthereumAddress): Promise<boolean> {
+    private async isStreamSubscriber_nonCached(streamId: StreamID, userId: Uint8Array): Promise<boolean> {
         try {
-            return await this.streamRegistryContractReadonly.hasPermission(streamId, userAddress, streamPermissionToSolidityType(StreamPermission.SUBSCRIBE))
+            const permission = await this.streamRegistryContractReadonly.getPermissionsForUserId(streamId, userId)
```
Comment (harbu) [inline packages/sdk/src/contracts/StreamRegistry.ts]: Consider renaming to 

`permissionsRecord`

File: packages/sdk/src/encryption/PublisherKeyExchange.ts [hunk 5]
```diff
@@ -42,7 +42,7 @@ export class PublisherKeyExchange {
     private readonly store: LocalGroupKeyStore
     private readonly authentication: Authentication
     private readonly logger: Logger
-    private readonly erc1271ContractAddresses = new Set<EthereumAddress>()
+    private readonly erc1271ContractAddresses = new ObjectSet<Uint8Array>((u) => binaryToHex(u, true))
```
Comment (harbu) [inline packages/sdk/src/encryption/PublisherKeyExchange.ts]: Move back to `Set<EthereumAddress>`. Follow-up question: do we need `ObjectSet` at all?

File: packages/sdk/src/permission.ts [hunk 6]
```diff
@@ -40,7 +39,7 @@ export const PUBLIC_PERMISSION_ADDRESS = '0x000000000000000000000000000000000000
 
 export type PermissionQueryResult = {
     id: string
-    userAddress: EthereumAddress
+    userAddress: string
```
Comment (harbu) [inline packages/sdk/src/permission.ts]: Should this be `userId`?

File: packages/sdk/test/integration/multiple-clients.test.ts [hunk 7]
```diff
@@ -73,14 +73,14 @@ describe('PubSub with multiple clients', () => {
...
+        const user = await client.getUserId()
         await stream.grantPermissions({ permissions: [StreamPermission.SUBSCRIBE], user })
         return client
     }
 
-    function checkMessages(published: Record<string, Message[]>, received: Record<string, MessageMetadata[]>) {
+    function checkMessages(published: Record<EthereumAddress, Message[]>, received: Record<EthereumAddress, MessageMetadata[]>) {
         for (const [key, msgs] of Object.entries(published)) {
-            expect(received[key].map((m) => m.signature)).toEqual(msgs.map((m) => m.signature))
+            expect(received[key as EthereumAddress].map((m) => m.signature)).toEqual(msgs.map((m) => m.signature))
```
Comment (harbu) [inline packages/sdk/test/integration/multiple-clients.test.ts]: Are these really `EthereumAddress` or just any hex string?

**Comment by teogeb** [comment]
Replaced with https://github.com/streamr-dev/network/pull/2774

**Pull request closed without merging**